riscv-debug-spec.tex

```diff
@@ -242,39 +259,80 @@ \subsubsection{Serial Ports}
 
 \section{Background}
 
-There are two forms of external debugging. The first is \emph{halt mode}
-debugging, where an external debugger will halt some or all components of a
-platform and inspect them while they are in stasis. Then the debugger can allow
-the hardware to either perform a single step or to run freely.
-
-The second is \emph{run mode} debugging.  In this mode there is a software
-debug agent running on a component (eg.  triggered by a timer interrupt on a
-RISC-V core) which communicates with a debugger without halting the component.
-This is essential if the component is controlling some real-time system (like a
-hard drive) where halting the component might lead to physical damage. It
-requires more software support (both on the chip as well as on the debug
-client).  For this use case the debug interface may include simple serial
-ports.
-
-A third use for the external debug interface is to use it as a general
-transport for a component to communicate with the outside world. For instance,
-it could be used to implement a serial interface that firmware could use to
-provide a simple CLI. This can use the same serial ports used for run-mode
-debugging.
+There are several use cases for dedicated debugging hardware, both
+internal to a CPU core and with an external connection.
+This specification defines techniques to support all of the use cases
+listed below. Some are
+optional to allow system designers to make cost vs capability tradeoffs.
+
+\begin{itemize}
+
+\item Debugging low-level software in the absence of an OS or other software.
+
+\item Debugging issues in the OS itself.
+
+\item Bootstrapping a system to test, configure, and program components before
+  there is any executable code path in the system.
+
+\item Accessing hardware on the system without a working CPU.
+
+\item Accesssing parts of a system which a CPU may not have software visibility
+  (e.g. cache tag arrays).
+
+\item Efficiently communicating data between a running process and external hardware,
+  for example to download code and data to write a program to memory.
```

Comment:
I wouldn't consider downloading code as an example of communicating between a running process and external hardware. You're blowing the running process away when you do that.